### PR TITLE
Chapter_3

### PR DESCRIPTION
**Pull request by stephanie1997-kek**

### CODE REVIEW

**Comment by stephanie1997-kek** [comment]
ready

File: app/assets/javascripts/static_pages.coffee [hunk 1]
```diff
@@ -0,0 +1,3 @@
+# Place all the behaviors and hooks related to the matching controller here.
+# All this logic will automatically be available in application.js.
+# You can use CoffeeScript in this file: http://coffeescript.org/
```
Comment (saobv-2223) [inline app/assets/javascripts/static_pages.coffee]: remove the file that never used

File: app/assets/stylesheets/static_pages.scss [hunk 2]
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the StaticPages controller here.
+// They will automatically be included in application.css.
+// You can use Sass (SCSS) here: http://sass-lang.com/
```
Comment (saobv-2223) [inline app/assets/stylesheets/static_pages.scss]: remove the file that never used

File: app/controllers/static_pages_controller.rb [hunk 3]
```diff
@@ -0,0 +1,7 @@
+class StaticPagesController < ApplicationController
+  def home
+  end
+
+  def help
+  end
```
Comment (saobv-2223) [inline app/controllers/static_pages_controller.rb]:
```suggestion
  def home; end

  def help; end
```

File: app/helpers/application_helper.rb [hunk 4]
```diff
@@ -1,2 +1,6 @@
 module ApplicationHelper
+	def full_title page_title = ""
+		base_title = "Ruby on Rails Tutorial Sample App"
+		page_title.empty? ? base_title : "#{page_title} | #{base_title}"
+	end
```
Comment (saobv-2223) [inline app/helpers/application_helper.rb]: Indent only 2 spaces
```suggestion
  def full_title page_title
    base_title = "Ruby on Rails Tutorial Sample App"
    page_title.blank? ? base_title : "#{page_title} | #{base_title}"
  end
```

File: app/helpers/static_pages_helper.rb [hunk 5]
```diff
@@ -0,0 +1,2 @@
+module StaticPagesHelper
```
Comment (saobv-2223) [inline app/helpers/static_pages_helper.rb]: remove the file that never used

File: app/views/layouts/application.html.erb [hunk 6]
```diff
@@ -1,14 +1,14 @@
 <!DOCTYPE html>
 <html>
   <head>
-    <title>SampleApp</title>
+    <title><%= full_title(yield(:title)) %></title>
```
Comment (saobv-2223) [inline app/views/layouts/application.html.erb]: remove () 
```suggestion
    <title><%= full_title yield(:title) %></title>
```

File: app/views/layouts/application.html.erb [hunk 7]
```diff
@@ -1,14 +1,14 @@
...
     <%= csrf_meta_tags %>
 
-    <%= stylesheet_link_tag    "application", media: "all", "data-turbolinks-track": "reload" %>
-    <%= javascript_include_tag "application", "data-turbolinks-track": "reload" %>
+    <%= stylesheet_link_tag    'application', media: 'all', 'data-turbolinks-track': 'reload' %>
+    <%= javascript_include_tag 'application', 'data-turbolinks-track': 'reload' %>
```
Comment (saobv-2223) [inline app/views/layouts/application.html.erb]: use ", please

File: app/views/layouts/application.html.erb [hunk 8]
```diff
@@ -1,14 +1,14 @@
...
   </head>
 
   <body>
-    <%= yield %>
+  	<%= yield %>
```
Comment (saobv-2223) [inline app/views/layouts/application.html.erb]:
```suggestion
    <%= yield %>
```

File: app/views/static_pages/help.html.erb [hunk 9]
```diff
@@ -0,0 +1,10 @@
+<% provide :title, "Help" %>
+
+<h1>Help</h1>
+<p>
+	Get help on the Ruby on Rails Tutorial at the
+	<a href="http://www.railstutorial.org/help">Rails Tutorial help page</a>.
+	To get help on this sample app, see the
+	<a href="http://www.railstutorial.org/book"><em>Ruby on Rails Tutorial</em>
+	book</a>.
```
Comment (saobv-2223) [inline app/views/static_pages/help.html.erb]: using 2 spaces, please

File: app/views/static_pages/help.html.erb [hunk 10]
```diff
@@ -0,0 +1,10 @@
...
+</p>
```
Comment (saobv-2223) [inline app/views/static_pages/help.html.erb]: add an empty line after line 10

File: app/views/static_pages/home.html.erb [hunk 11]
```diff
@@ -0,0 +1,8 @@
+<% provide :title, "Home" %>
+
+<h1>Welcome to the Sample App</h1>
+<p> 
+This is the home page for the
+<a href="http://www.railstutorial.org/">Ruby on Rails Tutorial</a>
+sample application.
```
Comment (saobv-2223) [inline app/views/static_pages/home.html.erb]: indenting 2 spaces and using link_to instead of a tag

File: app/views/static_pages/home.html.erb [hunk 12]
```diff
@@ -0,0 +1,8 @@
...
+</p>
```
Comment (saobv-2223) [inline app/views/static_pages/home.html.erb]: add an empty line after line 8

File: config/routes.rb [hunk 13]
```diff
@@ -1,3 +1,7 @@
 Rails.application.routes.draw do
+  root 'static_pages#home'
+
+  get 'static_pages/help'
+
```
Comment (saobv-2223) [inline config/routes.rb]: using " instead of ', remove line 5

File: test/controllers/static_pages_controller_test.rb [hunk 14]
```diff
@@ -0,0 +1,14 @@
+require 'test_helper'
+
```
Comment (saobv-2223) [inline test/controllers/static_pages_controller_test.rb]: We don't use test folder, so untrack them
1. Open `.gitignore` and add `test` line
2. Run `git rm --cached -r test`

File: app/assets/javascripts/static_pages.coffee (hunk 1 above)
Comment (stephanie1997-kek) [reply app/assets/javascripts/static_pages.coffee]: done

File: app/assets/stylesheets/static_pages.scss (hunk 2 above)
Comment (stephanie1997-kek) [reply app/assets/stylesheets/static_pages.scss]: done

File: app/controllers/static_pages_controller.rb (hunk 3 above)
Comment (stephanie1997-kek) [reply app/controllers/static_pages_controller.rb]: done

File: app/helpers/application_helper.rb (hunk 4 above)
Comment (stephanie1997-kek) [reply app/helpers/application_helper.rb]: done

File: app/helpers/static_pages_helper.rb (hunk 5 above)
Comment (stephanie1997-kek) [reply app/helpers/static_pages_helper.rb]: done

File: app/views/layouts/application.html.erb (hunk 6 above)
Comment (stephanie1997-kek) [reply app/views/layouts/application.html.erb]: done

File: app/views/layouts/application.html.erb (hunk 7 above)
Comment (stephanie1997-kek) [reply app/views/layouts/application.html.erb]: done

File: app/views/layouts/application.html.erb (hunk 8 above)
Comment (stephanie1997-kek) [reply app/views/layouts/application.html.erb]: done

File: app/views/static_pages/help.html.erb (hunk 9 above)
Comment (stephanie1997-kek) [reply app/views/static_pages/help.html.erb]: done

File: app/views/static_pages/help.html.erb (hunk 10 above)
Comment (stephanie1997-kek) [reply app/views/static_pages/help.html.erb]: done

File: app/views/static_pages/home.html.erb (hunk 11 above)
Comment (stephanie1997-kek) [reply app/views/static_pages/home.html.erb]: done

File: app/views/static_pages/home.html.erb (hunk 12 above)
Comment (stephanie1997-kek) [reply app/views/static_pages/home.html.erb]: done

File: config/routes.rb (hunk 13 above)
Comment (stephanie1997-kek) [reply config/routes.rb]: done

File: test/controllers/static_pages_controller_test.rb (hunk 14 above)
Comment (stephanie1997-kek) [reply test/controllers/static_pages_controller_test.rb]: done

**Comment by stephanie1997-kek** [comment]
ready

File: app/helpers/application_helper.rb [hunk 15]
```diff
@@ -1,2 +1,6 @@
 module ApplicationHelper
+  def full_title page_title = ""
+    base_title = "Ruby on Rails Tutorial Sample App"
+	page_title.empty? ? base_title : "#{page_title} | #{base_title}"
```
Comment (vanvtt-0952) [inline app/helpers/application_helper.rb]: 1. plz, remove code ` = ""`
2. Use I18n for `"Ruby on Rails Tutorial Sample App"` text
3. you should use `blank?`, don't use `empty?`

File: app/views/static_pages/help.html.erb [hunk 16]
```diff
@@ -0,0 +1,10 @@
+<% provide :title, "Help" %>
+
+<h1>Help</h1>
+<p>
+  Get help on the Ruby on Rails Tutorial at the
```
Comment (vanvtt-0952) [inline app/views/static_pages/help.html.erb]: Plz, using I18n, follow https://guides.rubyonrails.org/i18n.html

File: app/views/static_pages/help.html.erb [hunk 17]
```diff
@@ -0,0 +1,10 @@
...
+  <%= link_to "Rails Tutorial help page", "http://www.railstutorial.org/help" %>
+  To get help on this sample app, see the
+  <a href="http://www.railstutorial.org/book"><em>Ruby on Rails Tutorial</em>
```
Comment (vanvtt-0952) [inline app/views/static_pages/help.html.erb]: change `a` tag of html to use `link_to` tag

File: app/views/static_pages/home.html.erb (hunk 11 above)
Comment (vanvtt-0952) [reply app/views/static_pages/home.html.erb]: do you fix it yet?

Comment (stephanie1997-kek) [reply app/views/static_pages/home.html.erb]: I will do it now. Just now I updated it in wrong file, sorry.

**Comment by stephanie1997-kek** [comment]
ready

**Comment by stephanie1997-kek** [comment]
ready

File: app/controllers/application_controller.rb [hunk 18]
```diff
@@ -1,3 +1,18 @@
...
+  private
+
+  def switch_locale
+  	I18n.locale = extract_locale || I18n.default_locale
+  end
+
+  def extract_locale
+  	parsed_locale = params[:locale]
+  	I18n.available_locales.map(&:to_s).include?(parsed_locale) ? parsed_locale:nil
+  end
```
Comment (saobv-2223) [inline app/controllers/application_controller.rb]: indenting only 2 spaces, please

File: app/helpers/application_helper.rb [hunk 19]
```diff
@@ -1,2 +1,7 @@
 module ApplicationHelper
+  def full_title page_title
+
+    base_title = I18n.t "title"
+	page_title.blank? ? base_title : "#{page_title} | #{base_title}"
```
Comment (saobv-2223) [inline app/helpers/application_helper.rb]:
```suggestion
    base_title = I18n.t "title"
    page_title.blank? ? base_title : "#{page_title} | #{base_title}"
```

File: app/views/layouts/application.html.erb [hunk 20]
```diff
@@ -1,14 +1,14 @@
 <!DOCTYPE html>
 <html>
   <head>
-    <title>SampleApp</title>
+    <title><%= full_title yield(:title) %></title>
```
Comment (saobv-2223) [inline app/views/layouts/application.html.erb]: remove ()

File: app/views/layouts/application.html.erb [hunk 21]
```diff
@@ -1,14 +1,14 @@
...
     <%= csrf_meta_tags %>
 
     <%= stylesheet_link_tag    "application", media: "all", "data-turbolinks-track": "reload" %>
     <%= javascript_include_tag "application", "data-turbolinks-track": "reload" %>
   </head>
 
   <body>
-    <%= yield %>
+  <%= yield %>
```
Comment (saobv-2223) [inline app/views/layouts/application.html.erb]:
```suggestion
    <%= yield %>
```

File: config/locales/vi.yml [hunk 22]
```diff
@@ -0,0 +1,3 @@
+vi:
+  hello:
+  title:
```
Comment (saobv-2223) [inline config/locales/vi.yml]: indenting 2 spaces and add end line, please

File: config/locales/zh.yml [hunk 23]
```diff
@@ -0,0 +1,3 @@
+zh:
+  hello:
+  title:
```
Comment (saobv-2223) [inline config/locales/zh.yml]: same as above

File: config/routes.rb [hunk 24]
```diff
@@ -1,3 +1,10 @@
 Rails.application.routes.draw do
+#For detail on the DSL available within this file, see https://guides.rubyonrails.org/routing.html
+  root "static_pages#home"
+
+  get "static_pages/home"
+
+  get "static_pages/help"
+
```
Comment (saobv-2223) [inline config/routes.rb]:
```suggestion
  get "static_pages/home"
  get "static_pages/help"
```

**Review by saobv-2223 — COMMENTED** [review]
Open `.gitignore` file and add `test`
Then, run command `git rm --cached -r test` to untrack test folder

File: app/controllers/application_controller.rb (hunk 18 above)
Comment (stephanie1997-kek) [reply app/controllers/application_controller.rb]: done

File: app/helpers/application_helper.rb (hunk 19 above)
Comment (stephanie1997-kek) [reply app/helpers/application_helper.rb]: done

File: app/views/layouts/application.html.erb (hunk 21 above)
Comment (stephanie1997-kek) [reply app/views/layouts/application.html.erb]: done

File: config/locales/vi.yml (hunk 22 above)
Comment (stephanie1997-kek) [reply config/locales/vi.yml]: done

File: config/locales/zh.yml (hunk 23 above)
Comment (stephanie1997-kek) [reply config/locales/zh.yml]: done

File: config/routes.rb (hunk 24 above)
Comment (stephanie1997-kek) [reply config/routes.rb]: done

**Comment by stephanie1997-kek** [comment]
ready

File: app/controllers/application_controller.rb [hunk 25]
```diff
@@ -1,3 +1,18 @@
 class ApplicationController < ActionController::Base
-  protect_from_forgery with: :exception
+  before_action :switch_locale
+
+  def default_url_options
+  {locale: I18n.locale}
```
Comment (saobv-2223) [inline app/controllers/application_controller.rb]: indenting 2 spaces

File: app/controllers/application_controller.rb [hunk 26]
```diff
@@ -1,3 +1,18 @@
...
+  end
+
+  private
+
+  def switch_locale
+  I18n.locale = extract_locale || I18n.default_locale
```
Comment (saobv-2223) [inline app/controllers/application_controller.rb]: indenting 2 spaces

File: app/controllers/application_controller.rb [hunk 27]
```diff
@@ -1,3 +1,18 @@
...
+  end
+
+  def extract_locale
+  parsed_locale = params[:locale]
+  I18n.available_locales.map(&:to_s).include?(parsed_locale) ? parsed_locale:nil
```
Comment (saobv-2223) [inline app/controllers/application_controller.rb]: indenting 2 spaces

File: app/views/static_pages/help.html.erb [hunk 28]
```diff
@@ -0,0 +1,10 @@
+<% provide :title, t(".help_title") %>
+<h1><%= t ".heading" %></h1>
+
+<p>
+  <%= t ".description_title" %>
+  <%= link_to t("ruby_on_rails_help"), "http://www.railstutorial.org/help" %>
+  <%= t ".description_detail" %>
+  <%= link_to t("ruby_on_rails_book"), "http://www.railstutorial.org/book" %>
```
Comment (saobv-2223) [inline app/views/static_pages/help.html.erb]: with link such as `http://www.railstutorial.org/help` you should move it to settings.yml file and call this here.
You can research `gem config`

File: app/views/static_pages/home.html.erb [hunk 29]
```diff
@@ -0,0 +1,8 @@
+<% provide :title, t(".home_title") %>
+
+<h1><%= t ".heading" %></h1>
+<p> 
+  <%= t ".description_title" %>
+  <%= link_to t("ruby_on_rails_home"), "http://www.railstutorial.org/" %>
```
Comment (saobv-2223) [inline app/views/static_pages/home.html.erb]: same as above issue

**Review by saobv-2223 — COMMENTED** [review]
run command `git rm --cached -r test` to untrack test folder

**Comment by stephanie1997-kek** [comment]
> run command `git rm --cached -r test` to untrack test folder

image
I have solved this comment.

File: app/controllers/application_controller.rb (hunk 25 above)
Comment (stephanie1997-kek) [reply app/controllers/application_controller.rb]: done

File: app/controllers/application_controller.rb (hunk 26 above)
Comment (stephanie1997-kek) [reply app/controllers/application_controller.rb]: done

File: app/controllers/application_controller.rb (hunk 27 above)
Comment (stephanie1997-kek) [reply app/controllers/application_controller.rb]: done

File: app/views/static_pages/help.html.erb (hunk 28 above)
Comment (stephanie1997-kek) [reply app/views/static_pages/help.html.erb]: done

File: app/views/static_pages/home.html.erb (hunk 29 above)
Comment (stephanie1997-kek) [reply app/views/static_pages/home.html.erb]: done

**Comment by stephanie1997-kek** [comment]
ready

File: app/controllers/application_controller.rb (hunk 25 above)
Comment (chungpv-1008) [inline app/controllers/application_controller.rb]:
```suggestion
    {locale: I18n.locale}
```

File: app/controllers/application_controller.rb (hunk 26 above)
Comment (chungpv-1008) [inline app/controllers/application_controller.rb]:
```suggestion
    I18n.locale = extract_locale || I18n.default_locale
```

File: app/controllers/application_controller.rb (hunk 27 above)
Comment (chungpv-1008) [inline app/controllers/application_controller.rb]:
```suggestion
    parsed_locale = params[:locale]
    I18n.available_locales.map(&:to_s).include?(parsed_locale) ? parsed_locale : nil
```

File: config/routes.rb [hunk 30]
```diff
@@ -1,3 +1,10 @@
 Rails.application.routes.draw do
+#For detail on the DSL available within this file, see https://guides.rubyonrails.org/routing.html
+  scope "(:locale)", locale: /en/ do
+    root "static_pages#home"
+
+    get "static_pages/home"
+    get "static_pages/help"
+  end
   # For details on the DSL available within this file, see http://guides.rubyonrails.org/routing.html
```
Comment (chungpv-1008) [inline config/routes.rb]: remove all comments in this file
```suggestion
  scope "(:locale)", locale: /en/ do
    root "static_pages#home"

    get "static_pages/home"
    get "static_pages/help"
  end
```

File: config/settings.yml [hunk 31]
```diff
@@ -0,0 +1,5 @@
+links:
+  rails_tutorial:
+    home_page: http://www.railstutorial.org/
+    help_page: http://www.railstutorial.org/help
+    help_book: http://www.railstutorial.org/book
```
Comment (chungpv-1008) [inline config/settings.yml]: add end line (blank line) for this file

Comment (stephanie1997-kek) [reply config/settings.yml]: done

File: config/routes.rb (hunk 30 above)
Comment (stephanie1997-kek) [reply config/routes.rb]: done

File: app/controllers/application_controller.rb (hunk 27 above)
Comment (stephanie1997-kek) [reply app/controllers/application_controller.rb]: done

File: app/controllers/application_controller.rb (hunk 26 above)
Comment (stephanie1997-kek) [reply app/controllers/application_controller.rb]: done

File: app/controllers/application_controller.rb (hunk 25 above)
Comment (stephanie1997-kek) [reply app/controllers/application_controller.rb]: done

**Comment by stephanie1997-kek** [comment]
ready

File: app/controllers/application_controller.rb [hunk 32]
```diff
@@ -1,3 +1,18 @@
...
+
+  private
+
+  def switch_locale
+    I18n.locale = extract_locale || I18n.default_locale
+  end
+
+  def extract_locale
+    parsed_locale = params[:locale]
+    I18n.available_locales.map(&:to_s).include?(parsed_locale) ? parsed_locale:nil
```
Comment (chungpv-1008) [inline app/controllers/application_controller.rb]:
```suggestion
    I18n.available_locales.map(&:to_s).include?(parsed_locale) ? parsed_locale : I18n.default_locale
```

File: config/settings.yml [hunk 33]
```diff
@@ -0,0 +1,6 @@
...
+
```
Comment (chungpv-1008) [inline config/settings.yml]: add end line (blank line) for this file

Comment (stephanie1997-kek) [reply config/settings.yml]: image
I have add end line already.

File: app/controllers/application_controller.rb (hunk 32 above)
Comment (stephanie1997-kek) [reply app/controllers/application_controller.rb]: done

**Comment by stephanie1997-kek** [comment]
ready

**Review by saobv-2223 — COMMENTED** [review]
Do for test folder 
image

**Comment by stephanie1997-kek** [comment]
ready

**Comment by manhnt-0960** [comment]
merged

**Pull request merged**